apps/src/aichat/views/ChatEventsList.tsx

```diff
@@ -21,36 +23,98 @@ const ChatEventsList: React.FunctionComponent<ChatEventsListProps> = ({
   events,
   isTeacherView,
 }) => {
+  const [inProgrammaticScroll, setInProgrammaticScroll] = useState(false);
+  const [showScrollToBottom, setShowScrollToBottom] = useState(true);
   const {isWaitingForChatResponse} = useAppSelector(state => state.aichat);
 
   // Compare the chat events  as a string since the object reference will change on every update.
   // This way we will only scroll when the contents of the events have changed.
   const eventsString = JSON.stringify(events);
   const conversationContainerRef = useRef<HTMLDivElement>(null);
 
-  useEffect(() => {
+  const scrollToBottom = () => {
     if (conversationContainerRef.current) {
-      conversationContainerRef.current.scrollTo({
-        top: conversationContainerRef.current.scrollHeight,
-        behavior: 'smooth',
-      });
+      setShowScrollToBottom(false);
+
+      // On clear chat, we don't want to scroll to the bottom.
+      if (eventsString !== '[]') {
+        setInProgrammaticScroll(true);
+        conversationContainerRef.current.scrollTo({
+          top: conversationContainerRef.current.scrollHeight,
+          behavior: 'smooth',
+        });
+
+        const intervalId = setInterval(() => {
+          if (isAtBottom()) {
+            setInProgrammaticScroll(false);
+            clearInterval(intervalId);
+          }
+        }, 100);
+      }
+    }
+  };
+
+  const isAtBottom = () => {
+    const container = conversationContainerRef.current;
+
+    if (!container) {
+      return false;
     }
-  }, [eventsString, isWaitingForChatResponse]);
+
+    // Add a pixel of buffer to account for rounding errors.
+    return (
+      container.scrollTop + container.clientHeight + 1 >= container.scrollHeight
+    );
+  };
+
+  useEffect(() => {
+    const container = conversationContainerRef.current;
+
+    const handleScroll = () => {
+      if (container && !inProgrammaticScroll) {
+        setShowScrollToBottom(!isAtBottom());
+      }
+    };
+
+    container?.addEventListener('scroll', handleScroll);
+
+    return () => {
```

Comment:
Do you mind explaining this `return`? I haven't seen something like this before. We add an event listener to `scroll` event, and then we remove it shortly afterwards?

Reply:
Yup it's a "cleanup function", which runs a) before running the Effect on changed dependencies and b) when the component is unmounted, [more detail here](https://react.dev/reference/react/useEffect#parameters). Basically prevents us from continuously adding new event listeners when `inProgrammaticScroll` changes (and removes it on dismount, which I guess would happen when switching between levels?)

Reply:
On interesting - thanks!